---
title: "Calculating Genotype Probabilities"
teaching: 30
exercises: 30
---

:::::::::::::::::::::::::::::::::::::: questions 

- "How do I calculate QTL at positions between genotyped markers?"
- "How do I calculate QTL genotype probabilities?"
- "How do I calculate allele probabilities?"
- "How can I speed up calculations if I have a large data set?"

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

- To explain why the first step in QTL analysis is to calculate genotype 
probabilities.
- To calculate genotype probabilities.

::::::::::::::::::::::::::::::::::::::::::::::::

The first task in QTL analysis is to calculate conditional genotype 
probabilities, given the observed marker data, at each putative QTL position. 
For example, the first step would be to determine the probabilities for 
genotypes SS and SB at the locus indicated below.

![adapted from Broman & Sen, 2009](fig/unknown_genotype.png)

The `calc_genoprob()` function calculates QTL genotype probabilities, 
conditional on the available marker data. These are needed for most of the QTL 
mapping functions. The result is returned as a list of three-dimensional arrays 
(one per chromosome). Each 3d array of probabilities is arranged as individuals 
&times; genotypes &times; positions.

![Three-dimentional Genoporbs Array](fig/threeD_array.png){alt='Figure showing three-dimensional array of genoprobs'}

![See this page for a graphical review of data structures in R](http://venus.ifca.unican.es/Rintro/_images/dataStructuresNew.png).  

We'll use the
[Attie BL6/BTBR dataset](https://thejacksonlaboratory.box.com/shared/static/svw7ivp5hhmd7vb8fy26tc53h7r85wez.zip)
from 
[Tian et al](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4649649/)
(an intercross) as an example. In this study, circulating insulin levels were 
measured in an F2 cross between mouse strains C57BL/6J and BTBTR T+ <tf>. C57BL/6J 
mice exhibit low levels of non-heme iron, while SWR mice exhibit high levels. Iron 
levels between spleen and liver in the F2s were poorly correlated, indicating 
tissue-specific regulation. Significant QTL were found on chromosomes 2 and 16 
for liver, and on chromosomes 8 and 9 in spleen. Candidate genes included 
transferrin near the chromosome 9 peak, and <i>&beta;</i>2-microglobulin near 
the chromosome 2 peak.

First, we will load in the [qtl2]() library, which provides the functions that
we will use for QTL analysis.

```{r load_libraries}
suppressPackageStartupMessages(library(qtl2))
```


```{r load_data,eval=FALSE}
cross <- read_cross2(file = 'data/attie_b6btbr_grcm39/attie_control.json')
```

:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: instructor

We need the following block for the site to build on Github. The students do
not need to see or run the next block.

:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

```{r github_load_data,echo=FALSE,warning=FALSE,message=FALSE}
cross <- read_cross2(file = 'https://thejacksonlaboratory.box.com/shared/static/svw7ivp5hhmd7vb8fy26tc53h7r85wez.zip')
```


To load your own data from your machine, you would use the file path to your 
data files. For example, if the file path to your data files is 
`/Users/myUserName/qtlProject/data`, the command to load your data would look 
like this:

```{r load_my_data, eval=FALSE, error=FALSE}
myQTLdata <- read_cross2(file = "/Users/myUserName/qtlProject/data/myqtldata.json" )
```

The JSON file contains all control information for your data, including names of 
data files, cross type, column specifications for sex and cross information, and 
more. This can also be in YAML format. Alternatively, all data files can be 
zipped together for loading.

```{r load_my_zipdata, eval=FALSE, error=FALSE}
myQTLdata <- read_cross2(file = "/Users/myUserName/qtlProject/data/myqtldata.zip" )
```

Back to the iron data. Now look at a summary of the cross data and the names of 
each variable within the data.

```{r summary_data}
summary(cross)
names(cross)
```

Have a look at the markers listed in the genetic map, `gmap`. Markers are listed 
by chromosome and described by cM position. View only the markers on the first 
several chromosomes.

```{r map_data}
head(cross$gmap)
```

Next we use `calc_genoprob()` to calculate the QTL genotype probabilities.

```{r calc_genoprob}
probs <- calc_genoprob(cross      = cross, 
                       map        = cross$gmap, 
                       error_prob = 0.002)
```

The argument `error_prob` supplies an assumed genotyping error probability of 
0.002. If a value for `error_prob` is not supplied, the default probability is 
0.0001. 

Recall that the result of `calc_genoprob`, `probs`, is a list of three-dimensional 
arrays (one per chromosome). 

```{r list_chrs}
names(probs)
```

Each three-dimensional array of probabilities is arranged as individuals &times; 
genotypes &times; positions. Have a look at the names of each of the three 
dimensions for chromosome 19.

```{r view_array}
dimnames(probs$`19`)
```

View the first three rows of genotype probabilities for the first genotyped 
marker on chromosome 19, and the two adjacent pseudomarkers located at 1 cM 
intervals away. Compare the probabilities for each pseudomarker genotype with 
those of the genotyped marker.

```{r view_genoprob}
(probs$`19`)[1:5,,"rs4232073"]
```

We can also view the genotype probabilities using 
[plot_genoprob](https://github.com/rqtl/qtl2/blob/master/R/plot_genoprob.R). The 
arguments to this function specify:

1. probs: the genotype probabilities,
1. map: the marker map,
1. ind: the index of the individual to plot,
1. chr: the index of the chromosome to plot.

```{r plot_genoprob} 
plot_genoprob(probs = probs, 
              map   = cross$pmap, 
              ind   = 1, 
              chr   = 19, 
              main  = rownames(probs[['19']])[1])
```

The coordinates along chromosome 19 are shown on the horizontal axis and the 
three genotypes are shown on the vertical axis. Higher genotype probabilities 
are plotted in darker shades. This mouse has a RR genotype on the proximal end 
of the chromosome and transitions to BR.


::::::::::::::::::::::::::::::::::::: challenge 

## Challenge 1

<!-- DMG: Not sure about this. I'd rather plot more samples or chromosomes. -->

1). Load a second dataset from Arabidopsis recombinant inbred lines 
([Moore et al, Genetics, 2013](https://www.genetics.org/content/195/3/1077)) 
in a study of plant root response to gravity (gravitropism).  

`grav <- read_cross2(file = system.file('extdata', 'grav2.zip', package = 'qtl2'))`  
2). How many individuals were in the study? How many phenotypes? 
How many chromosomes?  
3). Insert pseudomarkers at 1cM intervals and save the results 
to an object called `gravmap`. Have a look at the first chromosome.  
4). Calculate genotype probabilities and save the results to an object 
called `gravpr`. View the genotypes for the first three markers and 
pseudomarkers on chromosome 1 for the first five individuals.   

:::::::::::::::::::::::: solution 

1). `grav <- read_cross2(file = system.file('extdata', 'grav2.zip', package = 'qtl2'))`  
2). `summary(grav)`   
3). `gravmap <- insert_pseudomarkers(map = grav$gmap, step = 1)`  
followed by `head(gravmap, n=1)`   
4). `gravpr  <- calc_genoprob(cross = grav, map = gravmap)` followed by  
`(gravpr$``1``)[1:5,,"PVV4"]`, `(gravpr$`1`)[1:5,,"c1.loc1"]`, and  
`(gravpr$`1`)[1:5,,"c1.loc2"]`

<!-- DMG: What about this? -->

```{r}
m  = maxmarg(probs)
ph = guess_phase(cross, m)
plot_onegeno(ph, cross$pmap)
```

:::::::::::::::::::::::::::::::::


## Challenge 3

Calculate genotype probabilities for a different data set from the 
[qtl2 data repository](https://github.com/rqtl/qtl2data), this one from a study 
of obesity and diabetes in a C57BL/6 (B6) Ã— BTBR intercross.   
1). Create a new script in RStudio with File -> New File -> R Script.  
2)  Download the B6 x BTBR zip file from the 
[qtl2 data repository](https://github.com/rqtl/qtl2data) into an object 
called `b6btbr` by running this code:  
`b6btbr <- read_cross2(file = "https://raw.githubusercontent.com/rqtl/qtl2data/master/B6BTBR/b6btbr.zip")`  
3). View a summary of the `b6btbr` data. How many individuals? phenotypes? 
chromosomes? markers?  
4). View the genetic map for the `b6btbr` data.  
5). Insert pseudomarkers at 2 cM intervals. Assign the results to an object 
called `b6btbrmap`.    
6). Calculate genotype probabilities assuming a genotyping error probability 
of 0.001. Assign the results to an object called `b6btbrpr`.    
7). View the first several rows of genotype probabilities for any marker on 
chromosome 18.  

:::::::::::::::::::::::: solution 

1). Create a new script in RStudio with File -> New File -> R Script.  
2). `b6btbr <- read_cross2(file = "https://raw.githubusercontent.com/rqtl/qtl2data/master/B6BTBR/b6btbr.zip`  
3). `summary(b6btbr)` shows 544 individuals, 3 phenotypes, 20 chromosomes, 
2057 markers.  
4). `b6btbr$gmap`  
5). `b6btbrmap <- insert_pseudomarkers(map=b6btbr$gmap, step=2)`  
6). `b6btbrpr <- calc_genoprob(cross=b6btbr, map=b6btbrmap, error_prob=0.001)`  
7). `dimnames((b6btbrpr$`18`))` shows all marker names for chromosome 18. 
`head((b6btbrpr$`18`)[,,"c18.loc48"])` gives genotype probabilities for an 
example pseudomarker, while `head((b6btbrpr$`18`)[,,"rs6338896"])`  gives 
genotype probabilities for a genotyped marker.

:::::::::::::::::::::::::::::::::


:::::::::::::::::::::::::::::::::::::


**Parallel calculations (optional)** To speed up the calculations with large 
datasets on a multi-core machine, you can use the argument `cores`. With 
`cores=0`, the number of available cores will be detected via 
`parallel::detectCores()`. Otherwise, specify the number of cores as a positive 
integer.

```{r calc_genoprob_multicore, eval=FALSE}
pr <- calc_genoprob(cross=iron, map=map, error_prob=0.002, cores=4)
```

**Allele probabilities (optional)** The genome scan functions use genotype 
probabilities as well as a matrix of phenotypes. If you wished to perform a 
genome scan via an additive allele model, you would first convert the genotype 
probabilities to allele probabilities, using the function 
`genoprob_to_alleleprob()`.

```{r allele_probs}
apr <- genoprob_to_alleleprob(probs = probs)
```

The figure below shows genotype and allele probabilities for 3 samples. In the 
Diversity Outbred, there are 36 possible genotype states 
(AA, AB, AC, ..., BB, BC, BD, ..., CC, CD, CE, ..., DD, DE, DF, ..., EE,...) or 
8 + 7 + 6 + 5 + 4 + 3 + 2 + 1. The first SNP below has genotype BB. In the table 
describing alleles (8 state founder probabilities), the probability that this 
SNP has a B allele is 1. The 2nd SNP has genotype BH, so the allele table shows 
a probability of 0.5 for B and 0.5 for H. The third SNP is either BG or BH, and 
has a probability of 0.5 for each of these genotypes. The allele table shows a 
probability of 0.5 for allele B, and 0.25 for both G and H.

![](../fig/geno-to-allele-probs.png)
::::::::::::::::::::::::::::::::::::: keypoints 

- The first step in QTL analysis is to calculate genotype probabilities.
- Calculate genotype probabilities between genotyped markers with 
calc_genoprob().

::::::::::::::::::::::::::::::::::::::::::::::::
